### PR TITLE
Allow copying beatmap link in song select carousel context menu

### PR DESCRIPTION
**Pull request by normalid-awa**
# Why
This could be very useful when playing on mobile, which player won't need to open their browser to just get the link. At the other side, desktop players can reduce the step of getting URLs, they don't need to jump to the browser and copy the link, instead just do it inside the game

Though the /np command can send the current beatmap to the chat, it doesn't make a use on if we want to share through the other app, like if we want to send it to discord.

# Implementation
Add `Copy link` item into both `DrawableCarouselBeatmapSet` and `DrawableCarouselBeatmap`
While `DrawableCarouselBeatmapSet` Copy URL as `$@"{api.WebsiteRootUrl}/beatmapsets/{beatmapSet.OnlineID}#{ruleset.ShortName}"` and `DrawableCarouselBeatmap` copy as `$@"{api.WebsiteRootUrl}/beatmapsets/{beatmapSet.OnlineID}#{ruleset.ShortName}/{beatmapInfo.OnlineID}"`

Screenshot 2024-08-21 at 11 37 17 AM

### CODE REVIEW

**Comment by Joehuu** [comment]
Just to list some other methods of doing the same thing in-game:
- `Details...` -> right click external link icon -> `Copy URL`
	- downsides: two extra steps and icon is potentially too small for mobile
- `/np` in chat, if the intention is sending the link there
	- downsides: the command is not known by many (probably would be noticed if there's some kind of UI when typing `/` like discord)

Note that the above only works while logged in, so that's probably another argument.

**Comment by normalid-awa** [comment]
> Just to list some other methods of doing the same thing in-game:
> 
> * `Details...` -> right click external link icon -> `Copy URL`
>   
>   * downsides: two extra steps and icon is potentially too small for mobile
> * `/np` in chat, if the intention is sending the link there
>   
>   * downsides: the command is not known by many (probably would be noticed if there's some kind of UI when typing `/` like discord)
> 
> Note that the above only works while logged in, so that's probably another argument.

Updated the description for better clarity

**Comment by bdach** [comment]
Normally I'd reject this based on the premise that you can already do this in other ways as @Joehuu said, but the number of reactions on the OP is giving me a little pause. I'm not sure whether people *didn't know* you could do this already (albeit, admittedly, in a more roundabout manner), or just *don't think that flow is good enough*.

**Comment by peppy** [comment]
I think this is handy to have here. Other methods are hidden or require more steps.

**Comment by peppy** [comment]
That said, this PR is basically a "needs full rewrite".

**Review by peppy — DISMISSED** [review]
Probably fine. Batch localisation should probably happen for context menus at some point.

File: osu.Game/Screens/Select/Carousel/DrawableCarouselBeatmap.cs [hunk 1]
```diff
@@ -288,6 +296,9 @@ public MenuItem[] ContextMenuItems
 
                 items.Add(new OsuMenuItem("Collections") { Items = collectionItems });
 
+                if (beatmapInfo.GetOnlineURL(api) is string url)
+                    items.Add(new OsuMenuItem("Copy link", MenuItemType.Standard, () => clipboard.SetText(url)));
```
Comment (Joehuu) [inline osu.Game/Screens/Select/Carousel/DrawableCarouselBeatmap.cs:300]: This doesn't have the OSD feedback like in `ExternalLinkButton`.

Comment (normalid-awa) [reply osu.Game/Screens/Select/Carousel/DrawableCarouselBeatmap.cs:300]: May I ask about what the OSD stands for?

Comment (peppy) [reply osu.Game/Screens/Select/Carousel/DrawableCarouselBeatmap.cs:300]: You can google for an explanation. But please leave this alone, I'll fix it to keep things moving.

Comment (Joehuu) [reply osu.Game/Screens/Select/Carousel/DrawableCarouselBeatmap.cs:300]: `OnScreenDisplay`, but I dunno if it's better to copy the logic from `ExternalLinkButton` two more times, or add another method to `OsuGame`.

Comment (peppy) [reply osu.Game/Screens/Select/Carousel/DrawableCarouselBeatmap.cs:300]: Latter is more valid. I believe we will be using this in many more places going forward.

File: osu.Game/Screens/Select/Carousel/DrawableCarouselBeatmap.cs [hunk 2]
```diff
@@ -288,6 +295,9 @@ public MenuItem[] ContextMenuItems
 
                 items.Add(new OsuMenuItem("Collections") { Items = collectionItems });
 
+                if (beatmapInfo.GetOnlineURL(api) is string url)
```
Comment (bdach) [inline osu.Game/Screens/Select/Carousel/DrawableCarouselBeatmap.cs:298]: No ruleset spec here? There is one on the beatmap set panels. And it does matter for stuff like converts.

Comment (peppy) [reply osu.Game/Screens/Select/Carousel/DrawableCarouselBeatmap.cs:298]: Yeah right, this was on my todo and I forgot to act on it.

**Review by bdach — APPROVED** [review]

**Pull request merged**